3.Linear_Modeling/nbconverted/fit_linear_model.py

```diff
@@ -0,0 +1,194 @@
+
+    # lm_results.head()
+    return lm_results
+
+
+def plot_lm(lm_df: pd.DataFrame, x: str, y: str, fill: str):
+    """Plot linear model beta coefficients
+
+    Parameters
+    ----------
+    lm_df : pd.DataFrame
+        Data Frame of outputed linear model values
+    x : str
+        x-axis column name
+    y : str
+        y-axis column name
+    fill : str
+        data-point fill column name
+    """
+
+    fig = px.scatter(lm_df, x=x, y=y, color=fill, title=f"Linear Model of {y}")
+    # fig.update_yaxes(range = [-0.01,0.5])
+    # fig.update_xaxes(range = [-2,2])
+    figure_file_path = pathlib.Path(
+        f"./figures/Linear_Model_{y}_beta_coefficient_against_{x}_beta_coefficient.html"
+    )
+    fig.write_html(figure_file_path)
+    fig.show()
+
+
+# In[3]:
+
+
+# Define inputs and outputs
+feature_file = pathlib.Path(
+    "../../Extracted_Features_(CSV_files)/interstellar_wave3_sc_norm_cellprofiler.csv.gz"
+)
+feature_df = pd.read_csv(feature_file, low_memory=False)
+output_dir = pathlib.Path("results")
+
+output_cp_file = pathlib.Path(output_dir, "linear_model_cp_features.tsv")
+
+
+# In[4]:
+
+
+pd.set_option("display.max_columns", 5000)
+
+
+# In[5]:
+
+
+print(feature_df.shape)
+feature_df.head()
+
+
+# In[6]:
+
+
+feature_df = feature_df.drop(feature_df.filter(regex="Costes").columns, axis=1)
+print(feature_df.shape)
+feature_df.head()
+
+
+# In[7]:
+
+
+feature_df = feature_df.replace(to_replace="/", value="_per_", regex=True)
+feature_df.head()
+
+
+# In[8]:
+
+
+# Recycled code from: https://github.com/WayScience/NF1_SchwannCell_data/blob/main/5_analyze_data/notebooks/linear_model/fit_linear_model.ipynb
+cell_count_df = (
+    feature_df.groupby("Metadata_Well")["Metadata_Plate"]
+    .count()
+    .reset_index()
+    .rename(columns={"Metadata_Plate": "Metadata_number_of_singlecells"})
+)
+
+feature_df = feature_df.merge(cell_count_df, on="Metadata_Well")
+
+# Drop na and reindex accordingly
+feature_df = (
+    feature_df.dropna()
+    .reset_index()
+    .assign(
+        Metadata_Treatment_and_Dose=lambda x: feature_df["Metadata_treatment"]
+        + "_"
+        + feature_df["Metadata_dose"]
+    )
+)
+
+cp_features = infer_cp_features(feature_df)
+print(f"We are testing {len(cp_features)} CellProfiler features")
+
+feature_df["Metadata_Treatment_and_Dose"].unique()
+
+
+# ##### Here I plot the beta coeifccents for each treatment against the number of cells per well. Data points the drift heavily in the Y axis are features that are affected the most by the y-axis treatment while data points that drift more in the x-axis are features that are most affected by the number of cells in a well.
+
+# In[9]:
+
+
+feature_df["Metadata_Treatment_and_Dose"].unique()
+
+
+# In[10]:
+
+
+for i in feature_df["Metadata_Treatment_and_Dose"].unique():
+    treatment = []
+    control = ["DMSO 0.1%_0"]
+    treatment.append(i)
+    lm_results = linear_model_function(control, treatment, feature_df, cp_features)
```

Comment:
```suggestion
control = ["DMSO 0.1%_0"]
for treatment in feature_df["Metadata_Treatment_and_Dose"].unique():
    lm_results = linear_model_function(control, treatment, feature_df, cp_features)
```

I think this does the same thing you have but in less steps

also note defining `control` prior to for loop

Reply:
oh! but your function must accept lists then (this is one downside of using a function at all! (the logic is split and in two places)).

can do `linear_model_function(control, treatment, [feature_df], cp_features)` instead